File: frontend/src/app/api/team/team.api.mapper.ts
import { TeamDto, TeamRegisterRequestDto, TeamUpdateRequestDto } from '@app/api/team/team.api.dto';
import { Team, TeamRegistration, TeamUpdate } from '@app/domain/team/team.model';

export class TeamApiMapper {

  static toDomain(dto: TeamDto): Team {
    return { ...dto };
  }

  static toRegisterRequest(registration: TeamRegistration): TeamRegisterRequestDto {
    return {
      name: registration.name,
      responsibleFirstName: registration.responsibleFirstName,
      responsibleLastName: registration.responsibleLastName,
      count: registration.count,
      paid: registration.paid,
    };
  }

  static toUpdateRequest(update: TeamUpdate): TeamUpdateRequestDto {
    return {
      name: update.name,
      responsibleFirstName: update.responsibleFirstName,
      responsibleLastName: update.responsibleLastName,
      paid: update.paid,
    };
  }
}

File: backend/src/main/java/abe/fvjc/tournament/api/team/TeamApiMapper.java
package abe.fvjc.tournament.api.team;

import abe.fvjc.tournament.domain.organisation.Person;
import abe.fvjc.tournament.domain.team.TeamRef;
import abe.fvjc.tournament.domain.team.TeamRegisterRequest;
import abe.fvjc.tournament.domain.team.TeamUpdateRequest;
import abe.fvjc.tournament.domain.team.TeamOverview;
import lombok.experimental.UtilityClass;

import java.util.List;

import static org.apache.commons.collections4.CollectionUtils.emptyIfNull;

@UtilityClass
public class TeamApiMapper {

    public static TeamRefDto toTeamRefDto(final TeamRef team) {
        return TeamRefDto.builder()
                .id(team.getId().value())
                .name(team.getName())
                .build();
    }

    static TeamDto toTeamDto(final TeamOverview overview) {
        return TeamDto.builder()
                .id(overview.getId().value())
                .name(overview.getName())
                .paid(overview.isPaid())
                .organisationId(overview.getOrganisationId().value())
                .responsibleFirstName(overview.getResponsible().getFirstName())
                .responsibleLastName(overview.getResponsible().getLastName())
                .build();
    }

    static List<TeamDto> toTeamDtos(final List<TeamOverview> overviews) {
        return emptyIfNull(overviews).stream()
                .map(TeamApiMapper::toTeamDto)
                .toList();
    }

    static TeamRegisterRequest toTeamRegisterRequest(final TeamRegisterRequestDto dto) {
        final var responsible = Person.builder()
                .firstName(dto.getResponsibleFirstName())
                .lastName(dto.getResponsibleLastName())
                .build();
        return TeamRegisterRequest.builder()
                .name(dto.getName())
                .responsible(responsible)
                .count(dto.getCount())
                .paid(dto.getPaid())
                .build();
    }

    static TeamUpdateRequest toTeamUpdateRequest(final TeamUpdateRequestDto dto) {
        final var responsible = Person.builder()
                .firstName(dto.getResponsibleFirstName())
                .lastName(dto.getResponsibleLastName())
                .build();
        return TeamUpdateRequest.builder()
                .name(dto.getName())
                .responsible(responsible)
                .paid(dto.getPaid())
                .build();
    }
}
